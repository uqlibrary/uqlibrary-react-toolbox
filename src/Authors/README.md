# Authors

Smart component that lists authors for UQ Library applications using ReactJs

## Props
- `dataSource: React.PropTypes.object.isRequired`
- `title: React.PropTypes.string`
- `addAuthor: React.PropTypes.func`
- `removeAuthor: React.PropTypes.func`
- `formValues: React.PropTypes.object`
- `loadAuthors: React.PropTypes.func.isRequired`
- `listOfAuthors: React.PropTypes.object.isRequired`
- `selectedAuthors: React.PropTypes.object`
- `form: React.PropTypes.string.isRequired.isRequired`
    
## Usage

**./App.js**
```jsx
import React from 'react';
...

import {Authors} from 'uqlibrary-react-toolbox';
export const authorDataSource = [
    {'id': 202, 'name': 'Author 1'},
    {'id': 263, 'name': 'Author 2'},
    {'id': 174, 'name': 'Author 3'},
    {'id': 177, 'name': 'Author 4'}
];
      
const App = () => (
  <div>
    <Authors 
        form="someFormName"
        dataSource={authorDataSource}
         />
  </div>
);

ReactDOM.render(
  <App />,
  document.getElementById('app')
);
```

