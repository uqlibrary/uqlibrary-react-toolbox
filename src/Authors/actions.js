// Repositories
import {getListOfAuthors} from '../../repositories/authors';

// Types
export const AUTHORS_LOADING = 'AUTHORS_LOADING';
export const AUTHORS_LOADED = 'AUTHORS_LOADED';
export const REMOVE_AUTHOR = 'REMOVE_AUTHOR';
export const ADD_AUTHOR = 'ADD_AUTHOR';

/**
 * Loads a list of authors
 * @returns {function(*)}
 */
export function loadAuthors() {
    return dispatch => {
        dispatch({type: AUTHORS_LOADING});
        getListOfAuthors().then(authorList => {
            dispatch({
                type: AUTHORS_LOADED,
                payload: authorList
            });
        }).catch((error) => {
            throw(error);
        });
    };
}

export function removeAuthor(authorIndex) {
    console.log('ai', authorIndex);
    return dispatch => {
        dispatch({
            type: REMOVE_AUTHOR,
            payload: authorIndex
        });
    };
}

export function addAuthor(authorID) {
    return dispatch => {
        dispatch({
            type: ADD_AUTHOR,
            payload: authorID
        });
    };
}
