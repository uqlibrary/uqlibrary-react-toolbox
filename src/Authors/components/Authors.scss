@import '../../../../../sass/variables';

.Author {
  padding: 0 0 10px 0;
  hr{
    margin: 0px 0 10px 0 !important;
  }
  .authorLeftIcon{
    padding: 3px;
    line-height: 32px !important;

  }
  .authorRightIcon {
    line-height: 32px;

  }
  .authorName {
    padding-left: 10px;
    line-height: 32px;
    font-weight: 500;
    text-overflow: ellipsis !important;
    white-space: nowrap !important;
    overflow: hidden !important;

    //border: 1px dashed red;
  }
}

// Padding between inputs when not in mobile <600px mode
div.row-sm.column.align-stretch.align-center-sm .inputPadding, div.row .inputPadding {
  padding-right: 16px;
}
@media (max-width: 600px){
  div.row-sm.column.align-stretch.align-center-sm .inputPadding {
    padding-right: 0px;
  }
}

// Set the MaterialUI inputs labels to ellipsis when too long..
.row-sm label, .row label, .column label, .column-sm label  {
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
  width: 100%;
}
