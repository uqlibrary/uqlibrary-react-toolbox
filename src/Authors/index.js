export {default as Authors} from './containers/Authors';

export {default as authorsReducer} from './reducer';
