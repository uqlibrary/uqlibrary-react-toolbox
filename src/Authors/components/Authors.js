import React, {Component} from 'react';
import {Field} from 'redux-form/immutable';

import Divider from 'material-ui/Divider';
import RaisedButton from 'material-ui/RaisedButton';

import {AutoCompleteSelect} from '../../AutoCompleteSelect';
import AuthorRow from './AuthorRow';

import './Authors.scss';

export default class Authors extends Component {

    static propTypes = {
        addAuthor: React.PropTypes.func,
        removeAuthor: React.PropTypes.func,
        formValues: React.PropTypes.object,
        loadAuthors: React.PropTypes.func.isRequired,
        listOfAuthors: React.PropTypes.object.isRequired,
        selectedAuthors: React.PropTypes.object,
        form: React.PropTypes.string.isRequired
    };

    constructor(props) {
        super(props);
    }

    componentDidMount = () => {
        this.props.loadAuthors();
    }

    addAuthor = () => {
        this.props.addAuthor(this.props.formValues.get('authorName'));
    }

    removeAuthor = (i) => {
        this.props.removeAuthor(i);
    }

    render() {
        const { listOfAuthors, formValues, selectedAuthors } = this.props;

        const ListOfAuthors = selectedAuthors.valueSeq().map((author, i) => {
            return(
                <AuthorRow key={i} authorID={author.get('id')} name={author.get('name')} removeAuthor={this.removeAuthor} />
            );
        });

        const authors = [];

        listOfAuthors.map((author) => {
            authors.push(
                {'id': author.get('id'), 'name': author.get('name')}
            );
        });

        return (
            <div>
                <div className="row" style={{marginBottom: '50px'}}>
                    <div className="flex inputPadding">
                        <Field component={AutoCompleteSelect} name="authorName"
                               maxSearchResults={10}
                               label="Author name (as published, in order)"
                               dataSource={authors}
                               dataSourceConfig={{text: 'name', value: 'id'}}
                               openOnFocus
                               fullWidth />
                    </div>
                    <div className="flex" style={{flex: '0 0 80px', textAlign: 'right'}}>
                        <RaisedButton label="Add" secondary style={{marginTop: '30px'}} onClick={this.addAuthor} disabled={formValues.size === 0} />
                    </div>
                </div>

                {ListOfAuthors}
                {ListOfAuthors.length > 0 &&
                    <Divider style={{margin: '10px 0 0 0'}}/>
                }
            </div>
        );
    }
}
