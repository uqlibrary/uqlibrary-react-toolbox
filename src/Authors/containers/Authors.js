import { connect } from 'react-redux';

import {reduxForm, getFormValues} from 'redux-form/immutable';
import Authors from '../components/Authors';
import Immutable from 'immutable';
import {loadAuthors, removeAuthor, addAuthor} from '../actions';


let AuthorsContainer = reduxForm({
    destroyOnUnmount: false
})(Authors);

AuthorsContainer = connect((state, initialProps) => {
    const listOfAuthorsState = state.get('authors');
    return {
        selectedAuthors: listOfAuthorsState.get('selectedAuthors'),
        listOfAuthors: listOfAuthorsState.get('listOfAuthors'),
        formValues: getFormValues(initialProps.form || 'AuthorForm')(state) || Immutable.Map({})
    };
}, dispatch => {
    return {
        loadAuthors: () => dispatch(loadAuthors()),
        removeAuthor: author => dispatch(removeAuthor(author)),
        addAuthor: author => dispatch(addAuthor(author))
    };
})(AuthorsContainer);

export default AuthorsContainer;
