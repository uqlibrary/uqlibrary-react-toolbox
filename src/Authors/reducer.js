import Immutable from 'immutable';

import { AUTHORS_LOADED, ADD_AUTHOR, REMOVE_AUTHOR } from './actions';

// Immutable state
const initialState = Immutable.fromJS({
    listOfAuthors: {},
    selectedAuthors: {}
});

const authorsReducer = (state = initialState, action) => {
    switch (action.type) {
        case AUTHORS_LOADED:
            return state.set('listOfAuthors', Immutable.fromJS(action.payload));
        case ADD_AUTHOR:
            const listOfAuthors = state.get('listOfAuthors');
            let updatedAuthorsList = Immutable.Set(state.get('selectedAuthors'));

            listOfAuthors.map(author => {
                if (author.includes(action.payload) && !updatedAuthorsList.has(author.get('id'))) {
                    const newAuthor = [Immutable.Map({'id': author.get('id'), 'name': author.get('name')})];
                    updatedAuthorsList = updatedAuthorsList.union(newAuthor);
                }
            });

            return state.set('selectedAuthors', Immutable.fromJS(updatedAuthorsList));
        case REMOVE_AUTHOR:
            const removeFromAuthorList = state.get('selectedAuthors').filter(author => {
                return author.get('id') !== action.payload;
            });
            return state.set('selectedAuthors', Immutable.fromJS(removeFromAuthorList));
        default:
            return state;
    }
};

export default authorsReducer;
