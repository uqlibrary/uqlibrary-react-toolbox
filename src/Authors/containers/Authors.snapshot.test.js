jest.dontMock('./Authors');

import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import Authors from './Authors';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';

const mockStore = configureStore([]);
const store = mockStore({});

function setup() {
    return shallow(<Provider store={store}><Authors form="aTestForm" /></Provider>);
}

describe('Authors snapshots tests', () => {
    it('renders authors field', () => {
        const wrapper = setup();
        const tree = toJson(wrapper);
        expect(tree).toMatchSnapshot();
    });
});
