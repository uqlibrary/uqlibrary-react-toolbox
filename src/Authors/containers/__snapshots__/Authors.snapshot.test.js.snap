// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Authors snapshots tests renders authors field 1`] = `
<Connect(ReduxForm)
  form="aTestForm"
/>
`;
